File: tasks-core-jvm/src/main/java/org/funfix/tasks/AsyncFun.java
package org.funfix.tasks;

import org.jspecify.annotations.NullMarked;

import java.io.Serializable;

/**
 * A function that is a delayed, asynchronous computation.
 */
@NullMarked
@FunctionalInterface
public interface AsyncFun<T> extends Serializable {
    Cancellable invoke(CompletionCallback<? super T> callback);
}

File: tasks-core/src/jvmMain/kotlin/org/funfix/tasks/AsyncFun.kt
package org.funfix.tasks

import org.jetbrains.annotations.NonBlocking
import java.io.Serializable

/**
 * A function that is a delayed, asynchronous computation.
 */
public fun interface AsyncFun<out T> : Serializable {
    @NonBlocking
    public operator fun invoke(
        executor: FiberExecutor,
        callback: CompletionCallback<T>
    ): Cancellable
}
